### ISSUE TITLE
Beacon node failing to recover after a SIGINT shutdown (fork choice store corruption)

### ISSUE THREAD

**Issue by jimmygchen** [sigp/lighthouse#4332]
## Description

A discord user reported that Lighthouse Beacon Node won't start after a `SIGINT` shutdown due to the below error. 

```
Failed to start beacon node, reason: Unable to load fork choice from disk: ForkChoiceError(ProtoArrayStringError("find_head failed: InvalidBestNode(InvalidBestNodeInfo...
```

Full beacon logs [here](https://seamonkey.tech/goerli/goerli-lighthouse-nethermind.lighthouse.log).

### Timelines

- 22:20:02 EE went offline
- 22:21:24 Starting to receive a lot of "Rejected gossip block" due to `ParentUnknown`
- 22:26:54 INFO Syncing, est_time: 7 mins, speed: 0.08 slots/sec, distance: 35 slots (7 mins)
- 22:56:54 INFO Syncing, est_time: --, distance: 185 slots (37 mins)
- 22:56:54 EE online, upcheck complete, state: Synced
- 22:56:59 User initiated SIGINT, seeing a bunch of beacon processing failures due to async tasks being cancelled.

### Diagnosis

It looks like the fork choice data got corrupted during the shutdown, and the beacon node was unable to recover from the invalid fork choice data stored on disk. We were able to obtain the fork choice DB dump from the user, which helped identity that `find_head` was failing due to the below evaluating to false:

https://github.com/sigp/lighthouse/blob/7f17b8e35127c0a082f04f0485b502ef933337e3/consensus/proto_array/src/proto_array.rs#L964-L965

After some digging with @michaelsproul, we think the root cause is likely due to the fork choice store mutation being aborted mid way, and in the shutdown handler we would persist the in memory fork choice store to disk, which could be corrupted due to the processing being half completed.

This could happen in various places, e.g. import blocks, pruning fork choice store after finalization etc.

I've placed `SIGINT` trigger in various places of the code but wasn't able to reproduce it consistently. I reproduced once by placing a `SIGINT` trigger at the end of the pruning function [here](https://github.com/sigp/lighthouse/blob/7f17b8e35127c0a082f04f0485b502ef933337e3/consensus/proto_array/src/proto_array.rs#L808), but wasn't able to reproduce this consistently.

### Additional Details from @michaelsproul 

> It's possible that the fork choice store's global checkpoints maybe have been updated before the block which did the justifying was added, here:

https://github.com/sigp/lighthouse/blob/c547a11b0da48db6fdd03bca2c6ce2448bbcc3a9/consensus/fork_choice/src/fork_choice.rs#L790-L810

> whereas the block itself doesn't get added to fork choice until here: 

https://github.com/sigp/lighthouse/blob/c547a11b0da48db6fdd03bca2c6ce2448bbcc3a9/consensus/fork_choice/src/fork_choice.rs#L865-L893

> if the thread running that function was cancelled midway, then the mutations to the fork choice store could be persisted, and stored on disk by the shutdown handler here:

https://github.com/sigp/lighthouse/blob/c547a11b0da48db6fdd03bca2c6ce2448bbcc3a9/beacon_node/beacon_chain/src/beacon_chain.rs#L5918

> so even though we're atomic with our writing of the fork choice store to disk, we are not atomic in memory when we account for threads getting cancelled

## Version

https://github.com/sigp/lighthouse/commit/981c72e863f37113f45306246775dcf715705a1c

## Present Behaviour

1. Start Lighthouse beacon node.
2. Terminate the process with `CTRL + C` or `SIGINT`.
3. Restart Lighthouse. Beacon node crashes with the following error.

```
CRIT Failed to start beacon node, reason: Unable to load fork choice from disk: ForkChoiceError(ProtoArrayStringError("find_head failed: InvalidBestNode(InvalidBestNodeInfo { current_slot: Slot(5683485), start_root: 0x5c44cef93338a459c34b62b641d1e11fd56a7d71e66e8e732a31723af91a800c, justified_checkpoint: Checkpoint { epoch: Epoch(177603), root: 0x5c44cef93338a459c34b62b641d1e11fd56a7d71e66e8e732a31723af91a800c }, finalized_checkpoint: Checkpoint { epoch: Epoch(177602), root: 0x1ebccdc8503c40e9c22467ffdbb6f752c6d502990739d6d84f40997d85ea7340 }, head_root: 0x5c44cef93338a459c34b62b641d1e11fd56a7d71e66e8e732a31723af91a800c, head_justified_checkpoint: Checkpoint { epoch: Epoch(177602), root: 0x1ebccdc8503c40e9c22467ffdbb6f752c6d502990739d6d84f40997d85ea7340 }, head_finalized_checkpoint: Checkpoint { epoch: Epoch(177601), root: 0x1453fc32d1cb2d476688857e7ca0c6487ce9f1bed5e36c027fc900194c7f346b } })")), module: lighthouse:588
```

## Expected Behaviour

Lighthouse beacon node should restart without crashing, even if the previous shutdown was triggered by a `SIGINT` signal (e.g. CTRL + C).

## Possible Solutions

A few options were being discussed:
1. Use a `copy-on-write` data structure for fork choice, the simplest version would be to replace the `HashMap` and `Vec`s in fork choice with types from here: https://docs.rs/rpds/latest/rpds/index.html
2. Keep tasks running until finish when after receiving a `SIGINT`, e.g. use `spawn_without_exit`. 
3. Remove fork choice persistence from shutdown handle

Option 1 is interesting and probably a good long term fix, but may be harder to do. 

Option 2 could be a bit complicated, because we use `spawn_async` in beacon processor, and it's probably something we'd want to avoid changing given the amount of surface it touches. 

Option 3 is probably the easiest and cleanest, we'd at most lose one epoch of data, and Lighthouse should be able to recover. Since Lighthouse usually recovers from `SIGKILL` OK, where the shutdown handlers were not triggered. Best to check with @paulhauner on this.

**Comment by paulhauner** [comment]
> thread running that function was cancelled midway

What are the scenarios where the thread is cancelled midway? My understanding is that this should only happen when there's an ungraceful termination (e.g. `kill -9`/`SIGKILL`). Specifically, it should not happen if we're giving it a `Ctrl + C`/`SIGINT`.

**Comment by jimmygchen** [comment]
```
May 21 22:56:59.776 INFO Shutting down.., reason: Success("Received SIGINT"), module: lighthouse:643
May 21 22:56:59.782 WARN Unable to free worker, error: channel closed, msg: did not free worker, shutdown may be underway, module: network::beacon_processor:2046
May 21 22:56:59.814 DEBG Persisting DHT to store, Number of peers: 259, service: network, module: network::service:967
May 21 22:56:59.821 INFO Saved DHT state, service: network, module: network::service:982
May 21 22:56:59.821 INFO Network service shutdown, service: network, module: network::service:991
May 21 22:56:59.891 DEBG Beacon block processing cancelled, error: BeaconChainError(TokioJoin(JoinError::Cancelled(Id(25890335)))), service: beacon, module: beacon_chain::beacon_chain:2713
May 21 22:56:59.892 WARN BlockProcessingFailure, outcome: TokioJoin(JoinError::Cancelled(Id(25890335))), msg: unexpected condition in processing block., module: network::beacon_processor::worker::sync_methods:555
May 21 22:56:59.892 DEBG Couldn't spawn task. Runtime shutting down, service: beacon, module: task_executor:294
May 21 22:56:59.892 ERRO Failed to spawn recompute head task, error: RuntimeShutdown, service: beacon, module: beacon_chain::canonical_head:546
May 21 22:56:59.892 DEBG Batch processing failed, service: sync, error: Internal error whilst processing block: TokioJoin(JoinError::Cancelled(Id(25890335))), imported_blocks: 21, last_block_slot: 5683360, chain: 16984880338948048269, first_block_slot: 5683297, batch_epoch: 177603, module: network::beacon_processor::worker::sync_methods:249
May 21 22:56:59.892 DEBG Could not send message to the sync service, error: channel closed, module: network::beacon_processor::worker:31
May 21 22:56:59.898 WARN Unable to free worker, error: channel closed, msg: did not free worker, shutdown may be underway, module: network::beacon_processor:2046
May 21 22:57:00.910 DEBG Couldn't spawn task. Runtime shutting down, service: beacon, module: task_executor:263
May 21 22:57:02.493 INFO Saved beacon chain to disk, service: beacon, module: beacon_chain::beacon_chain:5930
```

**Comment by michaelsproul** [comment]
The first log about cancellation comes from `BeaconChain::process_block`:

https://github.com/sigp/lighthouse/blob/c547a11b0da48db6fdd03bca2c6ce2448bbcc3a9/beacon_node/beacon_chain/src/beacon_chain.rs#L2712-L2719

The block processed comes from a chain segment (2nd log):

https://github.com/sigp/lighthouse/blob/c547a11b0da48db6fdd03bca2c6ce2448bbcc3a9/beacon_node/network/src/beacon_processor/worker/sync_methods.rs#L554-L559

And after the chain segment processing fails we try to re-run fork choice here:

https://github.com/sigp/lighthouse/blob/c547a11b0da48db6fdd03bca2c6ce2448bbcc3a9/beacon_node/network/src/beacon_processor/worker/sync_methods.rs#L368-L370

Which fails with this error (3rd log):

https://github.com/sigp/lighthouse/blob/c547a11b0da48db6fdd03bca2c6ce2448bbcc3a9/beacon_node/beacon_chain/src/canonical_head.rs#L545-L551

Fork choice is called from `BeaconChain::process_block` via `import_execution_pending_block`, which calls `import_block`, which calls `on_block`:

https://github.com/sigp/lighthouse/blob/c547a11b0da48db6fdd03bca2c6ce2448bbcc3a9/beacon_node/beacon_chain/src/beacon_chain.rs#L2683-L2694

https://github.com/sigp/lighthouse/blob/c547a11b0da48db6fdd03bca2c6ce2448bbcc3a9/beacon_node/beacon_chain/src/beacon_chain.rs#L2802-L2819

https://github.com/sigp/lighthouse/blob/c547a11b0da48db6fdd03bca2c6ce2448bbcc3a9/beacon_node/beacon_chain/src/beacon_chain.rs#L2897-L2908

The strange part is that the `import_block` happens inside a `spawn_blocking_handle`, and according to the [Tokio docs](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html) blocking tasks are _not interrupted by shutdown_:

> Closures spawned using spawn_blocking cannot be cancelled abruptly; there is no standard low level API to cause a thread to stop running.

> When you shut down the executor, it will wait indefinitely for all blocking operations to finish. You can use [shutdown_timeout](https://docs.rs/tokio/latest/tokio/runtime/struct.Runtime.html#method.shutdown_timeout) to stop waiting for them after a certain timeout. Be aware that this will still not cancel the tasks — they are simply allowed to keep running after the method returns. It is possible for a blocking task to be cancelled if it has not yet started running, but this is not guaranteed.

**Comment by michaelsproul** [comment]
Notably, _if_ Tokio were to return a JoinError in `spawn_blocking_handle`, we would log this debug log which we don't see:

https://github.com/sigp/lighthouse/blob/c547a11b0da48db6fdd03bca2c6ce2448bbcc3a9/common/task_executor/src/lib.rs#L298-L308

**Comment by michaelsproul** [comment]
Summary of current thinking:

- We don't think the corruption is caused by the SIGINT, per the reasoning above. It seems impossible for fork choice `on_block` to be cancelled mid-way (hat-tip to @paulhauner).
- Looking at Forky we found that the first block to update unrealized justification was 5683324. In the fork choice dump its unrealized justification is `None`, presumably because the block was imported by sync here with `CountUnrealized::False`:

https://github.com/sigp/lighthouse/blob/c547a11b0da48db6fdd03bca2c6ce2448bbcc3a9/beacon_node/network/src/sync/range_sync/chain.rs#L321-L325
- The _slot_ of the fork choice dump is 5683326 which matches the last block in the fork choice store. This also isn't consistent with a crash mid-way through `on_block`, because in that case we would expect the store's time and checkpoints to be updated ahead of its contained blocks (i.e. if we crashed importing 5683326 the time should be 5683326, but the last block should be earlier).
- There must be another bug lurking somewhere here that causes the fork choice store's global checkpoints to be updated _without_ corresponding blocks with those checkpoints being present in the store. This could either be because A) checkpoints are computed and used to update the global checkpoints which are not stored in the `ProtoNode`s, or B) checkpoints are computed and stored in the `ProtoNode`s but later blanked out. Attempts to find cases where either of these can happen have been unsuccessful. The only place we call `set_unrealized_justified_checkpoint` is on `on_block` where we _also set the node's unrealized checkpoint to `Some` (rules out A). Similarly, there don't seem to be any cases where we mutate the `unrealized_justified_checkpoint` of a `ProtoNode` (rules out B).

**Comment by michaelsproul** [comment]
New theory (untested):

One of the blocks with unrealized justification (i.e. 5683324/5/6) was imported _twice_: once via a chain segment, and once via RPC (or maybe gossip). In this case, a data race can occur between _checking whether the block is already known_ and _writing the block to fork choice_. The sequence of steps is:

1. While importing the chain segment, we check whether the block is already known using a read lock [here](https://github.com/sigp/lighthouse/blob/c547a11b0da48db6fdd03bca2c6ce2448bbcc3a9/beacon_node/beacon_chain/src/block_verification.rs#L1634-L1642).
2. Concurrently, the same block is being processed via RPC, and also does the same check using the read lock. It succeeds, because the block from the chain segment hasn't been added to fork choice yet.
3. The chain segment block obtains the _write_ lock on fork choice [here](https://github.com/sigp/lighthouse/blob/c547a11b0da48db6fdd03bca2c6ce2448bbcc3a9/beacon_node/beacon_chain/src/beacon_chain.rs#L2881-L2909), and calls `on_block`. Due to the block being processed with `CountUnrealized::False` this runs through `on_block` but _does not_ update the store's global `unrealized_justified_checkpoint`. A `ProtoNode` for the block gets written to the store containing `unrealized_justified_checkpoint: None`.
4. The RPC block obtains the same write lock in `import_block` and runs through `on_block` all over again. Because RPC blocks are imported with `CountUnrealized::True`, this time the store's global checkpoints _will_ be updated. However, when it comes to adding the `ProtoNode` containing the unrealized justification to the store, we will _return early without making any changes_ [here](https://github.com/sigp/lighthouse/blob/c547a11b0da48db6fdd03bca2c6ce2448bbcc3a9/consensus/proto_array/src/proto_array.rs#L370-L373).

The end result is a fork choice store that has had the global `unrealized_justified_checkpoint` updated, but contains no `ProtoNode` satisfying this. This is consistent with the fork choice dump we saw, which had an advanced global checkpoint and `ProtoNode::unrealized_justified_checkpoints` set to `Some` seemingly at random, i.e. in the cases where the RPC block was imported before the chain segment block (rather than the other way around).

In terms of a fix, it isn't sufficient to just remove the duplicate check from `ProtoArray::on_block`, because then the last block to be processed would just clobber the first one, and we'll get the same corruption if the RPC block is processed before the chain segment block.

I think the fundamental issue is that when we obtain the write lock we're acting on stale information from the previous time we held the read lock. This suggests that an appropriate fix might be to check that the block isn't already known right at the start of `ForkChoice::on_block`, i.e. [here](https://github.com/sigp/lighthouse/blob/c547a11b0da48db6fdd03bca2c6ce2448bbcc3a9/consensus/fork_choice/src/fork_choice.rs#L664-L667). In this case we can return `Ok(())` immediately and avoid the situation where the same block is applied to fork choice with different `CountUnrealized` behaviour.

Other options would be:

- Remove `CountUnrealized::False` altogether. We're currently using it to avoid extra work during sync, but with optimisations like #4118 this may not be so important.
- Hold a write lock from the first time we check for duplicates (too slow IMO).
- Make changes to sync to avoid processing blocks from RPC and chain segments simultaneously. This might be nice to do anyway to save work, but it feels like the wrong place to guard against this error. I think fixing it in fork choice would be preferable.

On the whole the impact of this bug is quite minimal, it can only cause damage if the node is shutdown while it's syncing, and _only if_ an RPC block is being processed concurrently.

**Comment by michaelsproul** [comment]
Resolved by #4357
